### PR TITLE
AFP, FTP: GUI to Middlewared

### PR DESCRIPTION
**Pull request by themylogin**
https://bugs.freenas.org/issues/25918
https://bugs.freenas.org/issues/25919
https://bugs.freenas.org/issues/26011

### CODE REVIEW

File: src/middlewared/middlewared/plugins/ftp.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+from middlewared.schema import accepts, Bool, Dict, Int, Str
+from middlewared.validators import Exact, IpAddress, Match, Or, Range
+from middlewared.service import SystemServiceService, ValidationErrors, CallError
+
+
+class FTPService(SystemServiceService):
+
+    service_name = "ftp"
+    key_prefix = "ftp_"
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/ftp.py]: Perhaps we can use the existing Config subclass here:

class Config:
   datastore_prefix = 'ftp_'

What do you think?

File: gui/middleware/form.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class MiddlewareModelForm:
+    key_prefix = NotImplemented
+    is_singletone = NotImplemented
+    middleware_plugin = NotImplemented
+
+    def clean(self):
+        try:
+            self.__update(dry_run=True)
```
Comment (william-gr) [inline gui/middleware/form.py]: I am curious about the reason for the dry run?

Doesn't all the validation occur before? Why would you need two calls to middleware?

Comment (themylogin) [reply gui/middleware/form.py]: Validation occurs when calling middleware `update` method from form's `save` method. Raising any validation errors there will lead to HTTP 500, so we should do validation earlier, in `clean` method as this is supposed to be with Django.

Comment (william-gr) [reply gui/middleware/form.py]: The freeadmin code path handles ValidationErrors from middleware client, returning a valid 200 json response so I am wondering why that can be used or why its a bad idea to do so.

Comment (themylogin) [reply gui/middleware/form.py]: I just did not knew about that code. I'll try to make use of it, thanks.

File: src/middlewared/middlewared/service.py [hunk 3]
```diff
@@ -206,6 +151,35 @@ def config(self):
         return await self.do_update(data)
 
 
+class SystemServiceService(ConfigService):
+    """
+    Service service abstract class
+
+    Meant for services that manage system services configuration.
+    """
```
Comment (william-gr) [inline src/middlewared/middlewared/service.py]: This service will be registered as a public service and available to users. Perhaps we need a way to identify classes as abstract?

e.g.
```
class Config:
    abstract = True
```
?

Comment (themylogin) [reply src/middlewared/middlewared/service.py]: Then all the child classes will need to redefine `abstract = False` to become public

We can solve this problem with decorator or metaclass approach, but I am fine with current solution: https://github.com/freenas/freenas/pull/328/commits/3ff019a5e540dd888568d803286bc93a43f0bd37

Comment (william-gr) [reply src/middlewared/middlewared/service.py]: Unless the default was `False` ;)
Config is not inherited.

But yeah, 3ff019a is fine.

**Pull request merged**